### PR TITLE
Add component for Balancer pool initialization

### PR DESCRIPTION
**Pull request by nlordell**
This PR adds a new abstraction for Balancer pool initialization. This will eventually be used by the pool fetcher in order to initialize its storage and avoid indexing all Balancer pool creation events to speed up initialization.

### Test Plan

Added new unit tests to check factory filtering logic.

### CODE REVIEW

File: shared/src/balancer/pool_init.rs [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+};
+use anyhow::{anyhow, Result};
+use contracts::{BalancerV2WeightedPool2TokensFactory, BalancerV2WeightedPoolFactory};
+use ethcontract::{Artifact, H160};
+
+#[derive(Debug, Default, PartialEq)]
+pub struct BalancerRegisteredPools {
+    pub weighted_pools: Vec<RegisteredWeightedPool>,
+    pub weighted_2token_pools: Vec<RegisteredWeightedPool>,
+}
```
Comment (bh2smith) [inline shared/src/balancer/pool_init.rs:18]: I assume this is only a temporary struct to be parsed and converted into `PoolStorage` when this is hooked up to the rest.

File: shared/src/balancer/pool_init.rs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+/// This can be used to index all pools from events instead of relying on the
+/// Balancer subgraph for example.
+pub struct EmptyPoolInitializer;
+
+#[async_trait::async_trait]
+impl PoolInitializing for EmptyPoolInitializer {
+    async fn initialize_pools(&self) -> Result<BalancerRegisteredPools> {
+        Ok(Default::default())
+    }
+}
```
Comment (bh2smith) [inline shared/src/balancer/pool_init.rs:36]: Doesn't it seem like overkill to make traits for this if we know we will only ever need it once? If it were going to be mocked in the test framework, then I see the point, but that doesn't seem to be the case.

**Review by bh2smith — APPROVED** [review]
Looks good, although I feel like some of the abstraction may be overkill for a single use component. Of course abstraction is fine and this is very good, perhaps for portability in the future.

File: shared/src/balancer/pool_init.rs (hunk 2 above)
Comment (nlordell) [reply shared/src/balancer/pool_init.rs:36]: This allows mocking and testing of how it interacts with the `BalancerPoolFetcher` once it gets hooked up. Additionally, it allows us to also have a `FilePoolInitializer` and `DatabasePoolInitializer` if we ever want to move away from the graph without having to refactor the calling code.

File: shared/src/balancer/pool_init.rs [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        // support (xDAI, Rinkeby and Mainnet) they are.
+        .ok_or_else(|| {
+            anyhow!(
+                "missing {} deployment for {}",
+                artifact.contract_name,
+                chain_id,
+            )
+        })?
+        .address)
+}
```
[the remark this answers is not in the document]
Comment (nlordell) [reply shared/src/balancer/pool_init.rs:117]: Hmm, from what I can tell, there is the generated method for getting a contract instance given a `Web3` instance:
```rust
        let weighted_pool_factory = BalancerV2WeightedPoolFactory::deployed(&web3).await?;
```

and a method for getting a block number from deployment information of a contract:
```
async fn get_deployment_block(
    deployment_info: Option<DeploymentInformation>,
    web3: &Web3,
) -> Option<u64>;
```

AFAICT there is no helper method for getting a contract address from a chain id in `event_handler` module.

**Review by vkgnosis — APPROVED** [review]

File: shared/src/balancer/pool_init.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bh2smith) [reply shared/src/balancer/pool_init.rs:117]: Right yes I was remembering her deployment block. I guess these are different so my comment doesn't matter.

**Pull request merged**